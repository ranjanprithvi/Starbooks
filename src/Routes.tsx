import React from "react";
import { Navigate, Route, Routes as ReactRoutes } from "react-router-dom";
import Books from "./components/Books/Books";
import BookDetails from "./components/Books/BookDetails";
import BookForm from "./components/Books/BookForm";
import NotFound from "./components/common/NotFound";
import Rentals from "./components/Rentals/Rentals";
import Users from "./components/Users/Users";
import RentalForm from "./components/Rentals/RentalForm";
import UserDetails from "./components/Users/UserDetails";
import UserForm from "./components/Users/UserForm";
import LoginForm from "./components/Users/LoginForm";
import ProtectedAdminComponent from "./components/common/ProtectedAdminComponent";
import ProtectedComponent from "./components/common/ProtectedComponent";
import AuthorGenreList from "./components/Authors/AuthorGenreList";
import Home from "./components/Home";

const Routes = ({ dataView }: { dataView?: string | undefined }) => {
    return (
        <ReactRoutes>
            <Route path="/" element={<Home />} />
            <Route path="/login" element={<LoginForm />} />
            <Route path="/books" element={<Books />} />
            <Route path="bookDetails/:id" element={<BookDetails />} />
            <Route
                path="books/:id"
                element={
                    <ProtectedAdminComponent>
                        <BookForm />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="users"
                element={
                    <ProtectedAdminComponent>
                        <Users dataView={dataView} />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="userDetails/:id"
                element={
                    <ProtectedComponent>
                        <UserDetails />
                    </ProtectedComponent>
                }
            />
            <Route
                path="users/:id"
                element={
                    <ProtectedAdminComponent>
                        <UserForm />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="rentals"
                element={
                    <ProtectedAdminComponent>
                        <Rentals />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="rentals/:id"
                element={
                    <ProtectedAdminComponent>
                        <RentalForm />
                    </ProtectedAdminComponent>
                }
            />
            <Route path="/authorGenreList" element={<AuthorGenreList />} />
            <Route path="/not-found" element={<NotFound />} />
            <Route path="*" element={<Navigate to="/not-found" replace />} />
        </ReactRoutes>
    );
};

export default Routes;
