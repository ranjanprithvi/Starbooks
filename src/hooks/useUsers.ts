import { User, UserQuery } from "../models/user";
import useData from "./generic/useData";

const useUsers = (query?: UserQuery, deps?: any[]) => {
    const {
        data: users,
        setData: setUsers,
        error,
        setError,
        isLoading,
    } = useData<User>("/users", query, deps);
    return { users, setUsers, error, setError, isLoading };
};

export default useUsers;
