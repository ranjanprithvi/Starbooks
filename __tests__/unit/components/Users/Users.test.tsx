import "@testing-library/jest-dom";
import React from "react";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Routes from "../../../../src/Routes";
import useUsers from "../../../../src/hooks/useUsers";

jest.mock("../../../../src/hooks/useUsers");
const mUseUsers = useUsers as jest.MockedFunction<typeof useUsers>;

Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
    })),
});

jest.mock("../../../../src/components/common/ProtectedAdminComponent");

beforeEach(() => {
    jest.clearAllMocks();
});

describe("Users", () => {
    it("renders correctly", () => {
        const tree = renderer
            .create(
                <MemoryRouter initialEntries={[`/users`]}>
                    <Routes />
                </MemoryRouter>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("displays the error in case of error", () => {
        mUseUsers.mockReturnValueOnce({
            users: [],
            isLoading: false,
            error: "Something went wrong",
            setUsers: jest.fn(),
            setError: jest.fn(),
        });

        render(
            <MemoryRouter initialEntries={[`/users`]}>
                <Routes />
            </MemoryRouter>
        );
        expect(screen.getByText(/something went wrong/i)).toBeInTheDocument();
    });
});
