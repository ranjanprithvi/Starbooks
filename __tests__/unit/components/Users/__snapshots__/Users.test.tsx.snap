// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users renders correctly 1`] = `
<div
  className="css-11dzgy"
>
  <div
    className="chakra-stack css-1lmrh1f"
  >
    <div
      className="chakra-stack css-15hkm2s"
    >
      <h2
        className="chakra-heading css-0"
      >
        Users
      </h2>
      <a
        className="chakra-button css-4xx2wk"
        href="/users/new"
        onClick={[Function]}
      >
        <span
          className="chakra-button__icon css-1wh2kri"
        >
          <svg
            aria-hidden={true}
            fill="currentColor"
            focusable={false}
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
            />
          </svg>
        </span>
        New User
      </a>
    </div>
    <hr
      aria-orientation="horizontal"
      className="chakra-divider css-1upb9tn"
    />
  </div>
</div>
`;
